### DISCUSSION TITLE
Parallelization across separate queries/systems

### DISCUSSION THREAD

**Discussion by LilithSilver** [genaray/Arch#145] [Ideas]
Sometimes, I want to run multiple queries or systems in parallel, not just a single query, and track dependencies between them.

A very mature version of this feature might look something like Unity's [Job Dependencies](https://docs.unity3d.com/Manual/JobSystemJobDependencies.html). Arch.Extended could build on top of that with something like Unity's [Dependency property](https://docs.unity3d.com/Packages/com.unity.entities@1.1/manual/scheduling-jobs-dependencies.html) which automatically tracks dependencies between parallel queries across systems and sets up the job dependencies accordingly.

But even without that magic dependency finding, any way of running multiple queries in parallel, even if handles would need to be tracked manually, would be appreciated! i.e. a simple example would be `var handle = world.ParallelQuery(...., DependencyJobHandle)` that returns a `JobHandle` itself for further scheduling. These chains could then be tracked across systems for the intrepid user (or by Arch.Extended of course).

Is something along these lines planned? Am I missing something that's already possible? I know parallelization isn't supported for sourcegen yet, but anything for regular queries?

Phenomenal work so far on this package, by the way. I've been really enjoying getting it running.

**Comment by genaray** [comment]
Thank you! Originally I had something like that even before and it is more or less still planned. The only problem is that multithreading is my biggest enemy. Every time I deal with the topic it brings me sooner or later to despair :D 

So if someone else could extend the JobScheduler I would be very grateful. The only problem is, it still has to be ZeroAlloc and as soon as dependencies come into play there is usually always some garbage somewhere.

**Reply by LilithSilver** [reply]
That makes sense. Do you think you could maybe make an issue or two and add it to the roadmap, since it's planned? It would be good to get that tracked. I might be able to contribute a little bit along those lines, but it would be a learning process for sure. 

Thanks for the reply!!

**Reply by genaray** [reply]
I added a issue for it! :)
